python/nav/ipdevpoll/shadows/__init__.py

```diff
@@ -260,6 +267,56 @@ def _fix_binary_garbage(self):
                 setattr(self, attr, repr(value))
         self.clear_cached_objects()
 
+    @classmethod
+    def prepare_for_save(cls, containers):
+        cls.changed_devices = cls._detect_version_changes(containers)
+        return super(Device, cls).prepare_for_save(containers)
+
+    @classmethod
+    def _detect_version_changes(cls, containers):
+        """
+        Detects if the software, hardware or firmware version changed for each device.
+
+        Returns a dictionary of the device and which of the versions changed.
+        """
+        if cls not in containers:
+            return
+
+        changed_devices = {}
+        for device in containers[cls].values():
+            old_device = device.get_existing_model()
+            if old_device:
+                changed_versions = list(
+                    set(device.get_diff_attrs(old_device)).intersection(
+                        (
+                            'hardware_version',
+                            'software_version',
+                            'firmware_version',
+                        )
+                    )
+                )
+                if changed_versions:
+                    changed_devices[device] = changed_versions
```

Comment:
I really like this implementation (although I don't think there is an explicit need to coerce the intersection into a `list` in line 289, unless you specifically intend to manipulate it as a list later) :)